Various (many...) interface and Reporter enhancements

loganharbour opened the pull request:
Most significant changes in PostprocessorInterface, VectorPostprocessorInterface, and UserObjectInterface:

- Makes various methods const
- Improves error checking; uses paramError where possible
- Restricts the checking of objects exist to after they are constructed
- Removes postprocessor metadata and instead does default value checking on-the-fly
- Adds dependency helpers to remove virtual overrides
- Adds error checking for all of the tests

This is ~1000 lines of tests/test objects, so it's _almost_ a reduction in framework code with lots of added functionality.

Refs #17512
Closes #17605

loganharbour committed:
Improve PostprocessorInterface

- Make const-correct where possible
- Use the related object's paramError and mooseError where possible
- Make Reporter more const-correct without the need for a write key
  for the PostprocessorInterface
- Improve the checking of the existance of a postprocessor
- Remove metadata store for postprocessors in InputParameters/Parser
  and instead check on the fly

Refs #17512

loganharbour committed:
Move UserObject methods to source/bottom of header

refs #17512

loganharbour committed:
Make interface methods that are overridden virtual

refs #17512

loganharbour committed:
Use UserObjectInterface methods with error checking

refs #17512

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/719171/) on 23d9924 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17527/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17527/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8b7394a261d838f23abc6744bf4cb16707c8ecf2`

aeslaughter commented on framework/include/ics/InitialConditionBase.h:
```diff
@@ -95,47 +95,48 @@ class InitialConditionBase : public MooseObject,
-
-  /// If set, UOs retrieved by this IC will not be executed before this IC
-  const bool _ignore_uo_dependency;
+  mutable std::set<std::string> _depend_uo;
```
I think this is the proper approach, but would add a detailed comment about why this is mutable.

aeslaughter commented on framework/include/interfaces/PostprocessorInterface.h:
```diff
@@ -74,29 +74,32 @@ class PostprocessorInterface
    * @see hasPostprocessorByName getPostprocessorValue
    */
-  bool hasPostprocessor(const std::string & name, unsigned int index = 0) const;
+  bool hasPostprocessor(const std::string & param_name, unsigned int index = 0) const;
```
Should you make the index const here as well?

aeslaughter commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -849,6 +849,12 @@ class FEProblemBase : public SubProblem, public Restartable
   ReporterData & getReporterData(ReporterData::WriteKey /*key*/) { return _reporter_data; }
 
+  template <typename T>
+  const T & getReporterValue(const ReporterName & reporter_name,
+                             const std::string & object_name,
+                             const ReporterMode & mode,
+                             const std::size_t time_index = 0) const;
```
Can we add some doco here, at least pointing to the ReporterData::getReporterData method. It took me a bit to remember that the "object_name" is the object asking and is simply for error reporting.

aeslaughter commented on framework/include/reporters/ReporterData.h:
```diff
@@ -292,12 +291,12 @@ class ReporterData
   /// Names of objects that have been declared
   std::set<ReporterName> _declare_names;
-  std::set<ReporterName> _get_names;
+  mutable std::set<ReporterName> _get_names;
```
Can you document the need for this as well.

loganharbour committed:
Add tests for PostprocessorInterface errors

refs #17512

loganharbour committed:
Add object information to paramError when the file location is not available

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/719359/) on d755876 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17527/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17527/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8b7394a261d838f23abc6744bf4cb16707c8ecf2`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/719421/) on bc41f3f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17527/site)

This comment will be updated on new commits.

loganharbour commented on framework/include/actions/ActionFactory.h:
```diff
@@ -125,6 +125,11 @@ class ActionFactory
+  /**
+   * Whether or not a task with the name \p task is registered.
+   */
+  bool isRegisteredTask(const std::string & task) const { return _tasks.count(task); }
```
Not at all. The std::size_t cast -> bool is in the standard, and is well defined

loganharbour committed:
Make sure function signatures are consistent

refs #17512

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/720480/) on 1f768b1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17527/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17527/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8b7394a261d838f23abc6744bf4cb16707c8ecf2`

loganharbour committed:
Improve VectorPostprocessorInterface

- Use the object's mooseError() and paramError() where possible
- Improve function signatures
- Use error checking in overridden methods

refs #17512

moosebuild commented:
Job [External App tests](https://civet.inl.gov/job/720533/) on 248f2bd : invalidated by @loganharbour

Moltres patched

loganharbour committed:
Check that in the case of non-insertion that values are the same

refs #17512

loganharbour committed:
Remove unnecessary reporter getter in FEProblemBase

refs #17512

loganharbour committed:
Remove unneeded hasPostprocesorObject method

refs #17512

loganharbour committed:
Add extra error checking when possible, add enabled pp method

refs #17512

loganharbour committed:
Add dependency helpers to PP, VPP, and UO interfaces

refs #17512

loganharbour committed:
Improve UserObjectInterface error testing

refs #17512

loganharbour committed:
Various fixups

refs #17512

loganharbour committed:
Add more tests for Reporter and ReporterInterface

refs #17512 #17605

loganharbour committed:
Reporter system enhancements

- Declare Reporter values for PPs and VPPs within the objects themselves
- Add type checking on declare/request for Reporters with the same name
- Add producer/consumer MooseObject ptrs in ReporterState
- Add getter for information about a Reporter (type, producers, consumers,
  context, modes)
- Add more error handling to ReporterData, Reporter, and ReporterInterface

refs #17512 #17605

loganharbour committed:
Add object for outputting Reporter debug information

refs #17605

zachmprince commented on framework/src/transfers/ReporterTransferInterface.C:
```diff
@@ -121,3 +139,44 @@ ReporterTransferInterface::getReporterNamesHelper(std::string prefix,
+                                 const FEProblemBase & from_problem,
+                                 const FEProblemBase & to_problem,
+                                 const std::string & method) const
+{
```
Could you just call the previous check function twice instead of having this one?

loganharbour replied:
👍🏼

zachmprince commented on framework/include/outputs/ReporterDebugOutput.h:
```diff
@@ -0,0 +1,30 @@
+protected:
+  /**
+   * Perform the debugging output
+   * For this object this is empty; the output is preformed in the constructor
```
This comment isn't true

zachmprince reviewed (changes requested):
Just some easy stuff I found. Looks really good to me.

loganharbour commented on framework/include/reporters/ReporterContext.h:
```diff
@@ -256,15 +255,16 @@ class ReporterContext : public ReporterContextBase
   /// Output data to JSON, see JSONOutput
   virtual void store(nlohmann::json & json) const override;
 
+  virtual std::string contextType() const override { return MooseUtils::prettyCppType<T>(); }
```
Ah, yep... this should be pure virtual

loganharbour commented on framework/include/restart/RestartableData.h:
```diff
@@ -145,7 +146,7 @@ template <typename T>
 RestartableData<T>::type()
 {
-  return typeid(T).name();
+  return MooseUtils::prettyCppType(this);
```
Oops. good catch

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/723741/) on 1480820 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17527/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17527/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 37a955d258e648ceb1743dabb5e88fc0ea3957b5`

loganharbour committed:
Make clang-format happy

loganharbour committed:
Add test design/doco pages for Reporter and VPP interfaces

refs #17605 #17512

loganharbour committed:
Re-add unused initPostprocessorData to be removed soon

refs #17512

loganharbour committed:
Move over-zealous assertion

refs #17605

loganharbour committed:
Clean up PP interface for default values and remove deprecated methods

refs #17512 #17605

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/725698/) on 75e9c50 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17527/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17527/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0c08243d3ae0bd6d64527bb2351f4a6c937d7c49`

loganharbour committed:
Re-add deprecated param that isn't ready to go yet

refs #17512

moosebuild commented:
Job [Controlled App tests](https://civet.inl.gov/job/726323/) on 6cbe6be : invalidated by @loganharbour

THM patched

aeslaughter commented on framework/include/interfaces/PostprocessorInterface.h:
```diff
@@ -76,29 +76,44 @@ class PostprocessorInterface
+   * @param index The index of the postprocessor
+   * @return The default value
+   */
+  PostprocessorValue getDefaultPostprocessorValue(const std::string & param_name,
```
Was dropping the `const &` intentional?

loganharbour replied:
Yeah. This isn't a reference for people to store. If they want the value at all, they should just call getPostprocessorValue(), which will create something they can bind to if it's a default value. I think that I should actually just remove this method. I don't think it's used.

aeslaughter replied:
I say remove it if you can.

aeslaughter commented on framework/include/reporters/ReporterContext.h:
```diff
@@ -417,9 +391,12 @@ class ReporterGeneralContext : public ReporterContext<T>
 
   virtual void resize(dof_id_type local_size) final;
+
+  virtual std::string contextType() const override { return MooseUtils::prettyCppType(this); }
```
Why can't this just be in ReporterContext?

loganharbour replied:
Because `this` changes depending on what context you have. If you put this just in `ReporterContext`, `this` will always be of type `ReporterContext<T>`, even if you call `ReporterGeneralContext<T>.contextType()`.

aeslaughter replied:
Ah yes, stupid C++. What about making this an input to the Constructor instead? It just seems silly to have the same method in each. I am thinking something like.

```
Context(..., const std::string & context_type = ""

NewContext : Context(..., MooseUtils::prettyCppType<NewContext>());

```

loganharbour replied:
This was how I had it originally. I'm fine with that. Will move

loganharbour replied:
Actually... this ends up adding quite a bit more code compared to just having the virtuals. I'm at a loss

aeslaughter commented on framework/src/postprocessors/Receiver.C:
```diff
@@ -29,18 +29,22 @@ Receiver::validParams()
   : GeneralPostprocessor(params),
     _initialize_old(getParam<bool>("initialize_old")),
-    _my_value(getPostprocessorValueByName(name())) // use FEProblem to avoid cyclic dependency
+    _my_value(_fe_problem.getReporterData().getReporterValue<PostprocessorValue>(
```
Why this change, seems like it should use the interface if it can.

loganharbour replied:
Yeah, I think you're right. Does the remaining stuff make sense, though? Needing write access to the data in order to request the older values. I'm trying to slowly wean off the use of the FEProblemBase calls (FEProblemBase::{getUserObject, getReporterValue, etc})

aeslaughter commented on framework/src/reporters/ReporterData.C:
```diff
@@ -15,138 +15,121 @@ ReporterData::ReporterData(MooseApp & moose_app) : _app(moose_app) {}
-    if (ptr->name().getObjectName() == object_name)
-    {
-      // Perform error checking for undeclared items
-      if (this->_declare_names.find(ptr->name()) == this->_declare_names.end())
-        mooseError("The Reporter value '", ptr->name(), "' was not declared.");
-
```
Where does this happen now? I am sad you are getting rid of my lambdas, I love these little guys.

loganharbour replied:
I think this should actually be an assertion now (didn't mean to remove it). It should be checked before hand in https://github.com/loganharbour/moose/blob/moar_moar_const/framework/src/reporters/ReporterData.C#L78

loganharbour committed:
Remove unused method

refs #17512

loganharbour committed:
Move Reporter producer ownership to the context

refs #17605

loganharbour commented:
The producer used to be owned by the ReporterState. It's now owned by the ReporterContext, which means that it's _required_ any time a context is defined. This makes more sense.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/727652/) on 931d472 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17527/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17527/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 911fe6223adc5d1a64e214b4762653eee33c380e`

aeslaughter commented on framework/src/reporters/ReporterData.C:
```diff
@@ -15,138 +15,157 @@ ReporterData::ReporterData(MooseApp & moose_app) : _app(moose_app) {}
+  if (context)
+  {
+    oss << context->getProducer().type() << " \"" << context->getProducer().name() << "\"";
+    oss << "\n  Context type:\n    " << context->contextType();
```
All that virtual stuff is all for this one method. Can we just used the C++ type stuff for this one piece of information?

loganharbour commented:
@aeslaughter Need yet another review due to a conflict after rebase. Hopefully can get this in today.
